pep-0718.rst

```diff
@@ -0,0 +1,152 @@
+PEP: 718
+Title: Subscriptable functions
+Author: James Hilton-Balfe <gobot1234yt@gmail.com>
+Sponsor: Guido van Rossum <guido@python.org>
+Discussions-To: https://discuss.python.org/t/26463/
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Content-Type: text/x-rst
+Created: 23-Jun-2023
+Python-Version: 3.13
+Post-History: `24-Jun-2023 <https://discuss.python.org/t/28457/>`__
+
+Abstract
+--------
+
+This PEP proposes making ``FunctionType``\ , ``MethodType``\ , ``BuiltinFunctionType``
+and ``BuiltinMethodType`` instances subscriptable for typing purposes. Doing so allows
+for specialisation in complex, hard-to-infer situations for type checkers where
```

Comment:
I'll update the abstract and motivation to include the top point more cause yeah that seems like a good idea. 

The point about being hard to infer is strictly about type checkers and them not being able to infer a concrete type for the type parameter rather than Unknown or Any (because it cannot be solved) and not about humans not being able to understand the intent. Sorry if that wasn't clear

Reply:
I think that is a much stronger emphasis.  The main point is that generic function specialization is implicit and type argument inference is unspecified (so it might not produce what a programmer expects, and checkers can differ on what they produce).  Explicit specialization allows a programmer to specify a type different from what inference might produce.  This also allows a programmer to write code that (should, in theory) typecheck the same with different checkers.